File: solutions/2.add-two-numbers/src/solution.rs
/*
 * @lc app=leetcode.cn id=2 lang=rust
 *
 * https://leetcode.cn/problems/add-two-numbers/
 */

pub struct Solution;

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

// @lc code=start
impl Solution {
    pub fn add_two_numbers(
        l_1: Option<Box<ListNode>>,
        l_2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        let mut dummy = ListNode::new(0);
        let mut p = &mut dummy;
        let mut p_1 = l_1;
        let mut p_2 = l_2;
        let mut carry = 0;
        while p_1.is_some() || p_2.is_some() || carry != 0 {
            let n_1 = match p_1 {
                Some(node) => {
                    p_1 = node.next;
                    node.val
                }
                None => 0,
            };
            let n_2 = match p_2 {
                Some(node) => {
                    p_2 = node.next;
                    node.val
                }
                None => 0,
            };
            let sum = n_1 + n_2 + carry;
            carry = sum / 10;
            p.next = Some(Box::new(ListNode::new(sum % 10)));
            p = p.next.as_mut().unwrap();
        }
        dummy.next
    }
}
// @lc code=end

File: solutions/2.add-two-numbers/src/solution.ts
/*
 * @lc app=leetcode.cn id=2 lang=typescript
 *
 * https://leetcode.cn/problems/add-two-numbers/
 */

/**
 * Definition for singly-linked list.
 */
class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val?: number, next?: ListNode | null) {
        this.val = val === undefined ? 0 : val;
        this.next = next === undefined ? null : next;
    }
}

// @lc code=start
function addTwoNumbers(
    l1: ListNode | null,
    l2: ListNode | null,
): ListNode | null {
    const dummy = new ListNode();
    let p = dummy;
    let p1 = l1;
    let p2 = l2;
    let carry = 0;
    while (p1 || p2 || carry) {
        const n1 = p1 ? p1.val : 0;
        p1 = p1 ? p1.next : null;
        const n2 = p2 ? p2.val : 0;
        p2 = p2 ? p2.next : null;
        const sum = n1 + n2 + carry;
        carry = Math.floor(sum / 10);
        p.next = new ListNode(sum % 10);
        p = p.next;
    }
    return dummy.next;
}
// @lc code=end

export { addTwoNumbers };
